drivers/exoscale/exoscale.go

```diff
@@ -462,20 +477,68 @@ func (d *Driver) Create() error {
 			ag = affinityGroup.ID
 		}
 		log.Debugf("Affinity group %v = %s", group, ag)
-		ags[idx] = ag
+		ags = append(ags, ag)
 	}
 
-	log.Infof("Generate an SSH keypair...")
+	// SSH key pair
 	keypairName := fmt.Sprintf("docker-machine-%s", d.MachineName)
-	kpresp, err := client.CreateKeypair(keypairName)
-	if err != nil {
-		return err
-	}
-	err = ioutil.WriteFile(d.GetSSHKeyPath(), []byte(kpresp.PrivateKey), 0600)
-	if err != nil {
-		return err
-	}
 	d.KeyPair = keypairName
+	if d.SSHKey == "" {
+		log.Infof("Generate an SSH keypair...")
+		resp, err := client.Request(&egoscale.CreateSSHKeyPair{
+			Name: keypairName,
+		})
+		if err != nil {
+			return fmt.Errorf("SSH Key pair creation failed %s", err)
+		}
+		keyPair := resp.(*egoscale.CreateSSHKeyPairResponse).KeyPair
+		if err = ioutil.WriteFile(d.GetSSHKeyPath(), []byte(keyPair.PrivateKey), 0600); err != nil {
+			return fmt.Errorf("SSH public key could not be written %s", err)
+		}
+	} else {
+		log.Infof("Importing SSH key from %s", d.SSHKey)
+
+		files := [][]string{
+			{d.SSHKey, d.GetSSHKeyPath()},
+			{d.SSHKey + ".pub", d.GetSSHKeyPath() + ".pub"},
+		}
+
+		for _, f := range files {
+			src := f[0]
+			dst := f[1]
+
+			if strings.HasPrefix(src, "~/") {
+				usr, _ := user.Current()
+				src = filepath.Join(usr.HomeDir, src[2:])
+			} else {
+				var err error
+				if src, err = filepath.Abs(src); err != nil {
+					return err
+				}
+			}
+
+			if err := mcnutils.CopyFile(src, dst); err != nil {
+				return fmt.Errorf("Unable to copy SSH file %s: %s", src, err)
+			}
+
+			if err := os.Chmod(dst, 0600); err != nil {
+				return fmt.Errorf("Unable to set permissions on the SSH file %s: %s", dst, err)
+			}
+		}
+
+		// Uploading public key
+		pubKey, err := ioutil.ReadFile(d.GetSSHKeyPath() + ".pub")
+		if err != nil {
+			return fmt.Errorf("Cannot read SSH public key %s", err)
+		}
+		_, err = client.Request(&egoscale.RegisterSSHKeyPair{
+			Name:      keypairName,
```

Comment:
Why not based the keypair name on the SSH key name instead to avoid uploading the same key multiple time?